compiler/codegen/OutOfLineCodeSection.cpp

```diff
@@ -198,3 +201,82 @@ TR::Node *TR_OutOfLineCodeSection::createOutOfLineCallNode(TR::Node *callNode, T
 
    return newCallNode;
    }
+
+void
+TR_OutOfLineCodeSection::assignRegisters(TR_RegisterKinds kindsToBeAssigned)
+   {
+   TR::Compilation *comp = _cg->comp();
+   if (hasBeenRegisterAssigned())
+      return;
+   // nested internal control flow assert:
+   _cg->setInternalControlFlowSafeNestingDepth(_cg->internalControlFlowNestingDepth());
+
+   // Create a dependency list on the first instruction in this stream that captures all current real register associations.
+   // This is necessary to get the register assigner back into its original state before the helper stream was processed.
+   //
+   _cg->incOutOfLineColdPathNestedDepth();
+
+   // this is to prevent the OOL entry label from resetting all register's startOfrange during RA
+   _cg->toggleIsInOOLSection();
+   TR::RegisterDependencyConditions *liveRealRegDeps = _cg->machine()->createCondForLiveAndSpilledGPRs(_cg->getSpilledRegisterList());
+
+   if (liveRealRegDeps)
+      {
+      _firstInstruction->setDependencyConditions(liveRealRegDeps);
+#if defined(TR_TARGET_ARM64)
+      liveRealRegDeps->bookKeepingRegisterUses(_firstInstruction, _cg);
+#endif
+      }
+   _cg->toggleIsInOOLSection(); // toggle it back because there is another toggle inside swapInstructionListsWithCompilation()
+
+
+   // Register assign the helper dispatch instructions.
+   swapInstructionListsWithCompilation();
+   _cg->doRegisterAssignment(kindsToBeAssigned);
+
+   TR::list<TR::Register*> *firstTimeLiveOOLRegisterList = _cg->getFirstTimeLiveOOLRegisterList();
+   TR::list<TR::Register*> *spilledRegisterList = _cg->getSpilledRegisterList();
+
+   for (auto li = firstTimeLiveOOLRegisterList->begin(); li != firstTimeLiveOOLRegisterList->end(); ++li)
+      {
+      if ((*li)->getBackingStorage())
+         {
+         (*li)->getBackingStorage()->setMaxSpillDepth(1);
+         traceMsg(comp,"Adding virtReg:%s from _firstTimeLiveOOLRegisterList to _spilledRegisterList \n", _cg->getDebug()->getName((*li)));
+         spilledRegisterList->push_front((*li));
+         }
+      }
```

Comment:
This isn't done on ARM and P, might be worth guarding appropriately.

Reply:
Also, FWIW, this isn't seen in the x implementation either.

Reply:
These actually need to be used on all platforms. They just haven't encountered the bug yet. It's a similar bug that can occur as described in #5977. The code is unguarded in anticipation of fixing this on all platforms as we continue the unification of the register allocator.